Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/certificate/DccV1Parser.kt

```diff
@@ -0,0 +1,136 @@
+package de.rki.coronawarnapp.covidcertificate.common.certificate
+
+import com.google.gson.Gson
+import com.upokecenter.cbor.CBORObject
+import dagger.Reusable
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidVaccinationCertificateException
+import de.rki.coronawarnapp.util.serialization.BaseGson
+import de.rki.coronawarnapp.util.serialization.fromJson
+import timber.log.Timber
+import javax.inject.Inject
+
+@Reusable
+class DccV1Parser @Inject constructor(
+    @BaseGson private val gson: Gson
+) {
+    fun parse(map: CBORObject, mode: Mode): DccV1 = try {
+        map[keyHCert]?.run {
+            this[keyEuDgcV1]?.run {
+                this.toCertificate().toValidated(mode)
+            } ?: throw InvalidVaccinationCertificateException(ErrorCode.HC_CWT_NO_DGC)
+        } ?: throw InvalidVaccinationCertificateException(ErrorCode.HC_CWT_NO_HCERT)
+    } catch (e: InvalidHealthCertificateException) {
+        throw e
+    } catch (e: Throwable) {
+        throw InvalidHealthCertificateException(ErrorCode.HC_CBOR_DECODING_FAILED, cause = e)
+    }
+
+    private fun CBORObject.toCertificate(): DccV1 = try {
+        val json = ToJSONString()
+        gson.fromJson(json)
+    } catch (e: InvalidHealthCertificateException) {
+        throw e
+    } catch (e: Throwable) {
+        throw InvalidHealthCertificateException(ErrorCode.JSON_SCHEMA_INVALID)
+    }
+
+    private fun DccV1.toValidated(mode: Mode): DccV1 = try {
+        this.run {
```

Comment:
Would this do the same?

```kotlin
           checkModeRestrictions(mode).apply {
                // Apply otherwise we risk accidentally accessing the original obj in the outer scope
                require(isSingleCertificate())
                checkFields()
            }
```